### PR TITLE
Refactor config

### PR DESCRIPTION
**Pull request by leotrs**
This is one of a series of PRs aimed at fixing #337. This PR does a few things:

1. Introduces the `tempconfig` context manager. This works as follows:
    ```python
       c = Camera()
       c.frame_width == config['frame_width']        # -> True
       with tempconfig({'frame_width': 100}):
           c = Camera()
           c.frame_width == config['frame_width']    # -> False
           c.frame_width == 100                      # -> True
    ```
    So it temporarily changes the config so that all classes that use it for default values will use the new (temporary) config. After 
    the `with` statement ends, the config is restored. This accounts for the diff in `config.py`, `__init__`, and `test_config.py`. I remember @PgBiel in the past requested the feature of changing the global `config`. Now we can do it!

2. The problem with the code above is that `Camera` instances used the value of `frame_width` that was available at the time when the `Camera` class was being defined, not the value of the `config` at the time that the instance was being constructed. So I had to remove a few keys from `Camera.CONFIG` and set them in `Camera.__init__` instead. I did it in a way that it will later be easy to change to `attrs` cc @PgBiel. This accounts for the diff in `camera.py` and `test_camera.py`

3. For some reason my changes made a difference in `test_logging.py`. I made it so that this test doesn't care about whitespace in the logs and now it works. I believe this is a fair change for now since apparently #335 is going to deprecate how tests currently work. cc @huguesdevimeux 

Hope this makes sense! This is just the first in a few PRs concerning the config system.

### CODE REVIEW

**Comment by huguesdevimeux** [comment]
> For some reason my changes made a difference in test_logging.py. I made it so that this test doesn't care about whitespace in the logs and now it works. I believe this is a fair change for now since apparently #335 is going to deprecate how tests currently work. 

I think @Aathish04 is better to fix that, as he wrote this test.

File: manim/camera/camera.py
Comment (huguesdevimeux) [inline manim/camera/camera.py:80]: **Question** : So, if I understand correctly, you said that now this uses `kwargs` in addition to the config dict for these attributes, doesn't it? 
If yes, I think you should add a quick line that would explain that, because your comment is very good but ... I didn't get it immediately.

File: tests/test_logging/test_logging.py
Comment (huguesdevimeux) [inline tests/test_logging/test_logging.py:57]: **Issue :** I _think_ it's not the right way to fix these tests. See my review comment.

**Review by huguesdevimeux — CHANGES_REQUESTED** [review]
Your test fix is I think not good. 
I think it's because the log_width parameters in the config is not right, as it should be set to 512. 
One way to fix it would be to create a custom `.cfg` with `log_width = 512` and to use it in the CLI in the logging test. That's what I did in #335 

cc @Aathish04 (author of this test)

File: tests/test_logging/test_logging.py
Comment (leotrs) [reply tests/test_logging/test_logging.py:57]: As far as I can tell, I have to add a `manim.cfg` file to the directory where `test_logging.py` lives, correct?

If that's the case, I'm honestly far more inclined to change the test than add that file. In fact, the whole point of this PR (and more that will come after it) is so that _we explicitly don't have to do that_. So if I add the `manim.cfg` file here, it will just go away (very) soon.

File: manim/camera/camera.py
Comment (leotrs) [reply manim/camera/camera.py:80]: Done in latest commit

File: tests/test_logging/test_logging.py
Comment (huguesdevimeux) [reply tests/test_logging/test_logging.py:57]: The issue is that this test run a CLI invocation, so the manim.cfg have to be there, no ?

Comment (Aathish04) [reply tests/test_logging/test_logging.py:57]: I was under the impression that all the tests will be run with one "main" custom `manim.cfg` file, like the "user-wide" cfg file, and those tests which _absolutely_ need to have different cfg params, like the test that checks whether the cfg files cascade properly, will have folder-wide cfg files.

If my understanding is correct, we can simply set the log width to some large number, say 1024 or 2048, in the tests-wide config, and still keep our character-per-character checks for the log testing.

Comment (huguesdevimeux) [reply tests/test_logging/test_logging.py:57]: Yes, we can do that as well.

Btw, for clarity purpose I used a folder wide cfg file in #335, could you check it out ?

Comment (leotrs) [reply tests/test_logging/test_logging.py:57]: > If my understanding is correct, we can simply set the log width to some large number, say 1024 or 2048, in the tests-wide config, and still keep our character-per-character checks for the log testing.

This sounds like the way to go. But also, this shouldn't be part of this PR, no? Also if I understand correctly, all of the logs will have to be regenerated so that they match the new width. Yes?  Would it be OK with you two if I did all of that in a separate PR that would (hopefully) be merged quickly?

Comment (huguesdevimeux) [reply tests/test_logging/test_logging.py:57]: Yes, you're right. Just fix this here, make tests pass. 

I'm convinced that we should change the way to test logs, and so we will use a JSON format in the future. See #365

Comment (leotrs) [reply tests/test_logging/test_logging.py:57]: > Just fix this here, make tests pass.

I'm a little slow today. Do you mean that we should keep the `replace` call in this PR, merge it and then prepare another PR fixing the log width thing? If that's what you meant, then this PR is ready to be merged.

Comment (huguesdevimeux) [reply tests/test_logging/test_logging.py:57]: If tests are working, then I'm all good.

Comment (Aathish04) [reply tests/test_logging/test_logging.py:57]: @leotrs @huguesdevimeux  I found out the exact reason for the expected and actual logs not matching.

`rich`, after every logger.log() call it caters to, prints the line number of the log call, like so:

`DEBUG     Read configuration files:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 config.py:  42`

Since `config.py` has been refactored, that _particular_ `logger.log()` call has been shifted from a 2 digit line number to a 3 digit line number, causing an imbalance in the whitespace, since the regex replaces all digits with whitespace since they change.

The tests won't pass if you increase the log width, they'll pass only if you change the expected log to account for this, kind of like regenerating the reference data for the graphical tests.

I would suggest replacing the current expected logs file with the log file it actually makes.

**Review by huguesdevimeux — APPROVED** [review]

File: tests/test_logging/test_logging.py
Comment (huguesdevimeux) [reply tests/test_logging/test_logging.py:57]: > I would suggest replacing the current expected logs file with the log file it actually makes.

Let's listen to master test logging :D

Comment (leotrs) [reply tests/test_logging/test_logging.py]: @Aathish04 you are _brilliant_. Just replaced the file, and as soon as the checks are done I'm merging this

**Comment by leotrs** [comment]
Ok, fixed the tests thanks to @Aathish04.

Now we need one more approval.

**Review by Aathish04 — APPROVED** [review]

**Pull request merged**
